File: 2015/day1b.c
#include <stdio.h>

int main(int argc, char **argv) {
	char * dir;
	int floor = 0;
	int no_basement = 1;
	if (argc != 2) {
		printf ("Please provide your puzzle input as the first argument\n");
		return 1;
	}
	dir = argv[1];
	while (*dir) {
		switch (*dir++) {
			case '(':
				floor++;
				break;
			case ')':
				floor--;
		}
		if ((floor < 0) && no_basement) {
			printf ("Santa arrived in the basement at position %d\n", (int) (dir-argv[1]));
			no_basement = 0;
		}
	}
	printf ("Santa should arrive on floor %d\n", floor);
}

File: 2015/day1b.go
package main

import (
	"fmt"
	"os"
	"log"
	"io/ioutil"
)

func printusage() {
	fmt.Printf("Usage: %s <directions file>\n\nFind Santa's location based on up and down directions given in <directions file>\n", os.Args[0])
	log.Fatal("FATAL: no filename given")
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	if len(os.Args) != 2 {
		printusage()
	}

	dat, err := ioutil.ReadFile(os.Args[1])
	check(err)

	floor := 0
	basement_reached := false
	for i := 0; i < len(dat); i++ {
		switch(dat[i]) {
			case '(':
				floor++
			case ')':
				floor--
		}
		if !basement_reached && (floor < 0) {
			fmt.Printf("Santa reached the basement at position %v\n", i+1)
			basement_reached = true
		}
	}
	fmt.Printf("Santa should arrive on floor %v\n", floor)
}
